stock_move_change_source_location/wizards/stock_move_change_source_location_wizard.py

```diff
@@ -0,0 +1,119 @@
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class StockMoveChangeSourceLocation(models.TransientModel):
+    _name = "stock.move.change.source.location.wizard"
+    _description = "Stock Move Change Source Location Wizard"
+
+    def _prepare_default_values(self, picking):
+        return {"location_id": picking.location_id.id}
+
+    @api.model
+    def default_get(self, fields):
+        res = super().default_get(fields)
+        active_model = self.env.context["active_model"]
+        active_ids = self.env.context["active_ids"] or []
+        picking = self.env[active_model].browse(active_ids)
+        res.update(self._prepare_default_values(picking))
+        return res
+
+    def _default_old_location_id(self):
+        stock_picking_obj = self.env["stock.picking"]
+        pickings = stock_picking_obj.browse(self.env.context["active_ids"])
+        first_move_line = pickings.mapped("move_line_ids")[:1]
+        return first_move_line.location_id.id
+
+    def _get_allowed_locations(self):
+        return ["internal"]
+
+    def _get_allowed_location_domain(self):
+        return [("usage", "in", self._get_allowed_locations())]
+
+    def _get_allowed_picking_states(self):
+        return ["confirmed", "assigned"]
+
+    location_id = fields.Many2one(
+        comodel_name="stock.location",
+        string="Actual location",
+        required=True,
+        readonly=True,
+    )
+    old_location_id = fields.Many2one(
+        comodel_name="stock.location",
+        string="Old location",
+        default=_default_old_location_id,
+        domain=lambda self: self._get_allowed_location_domain(),
+    )
+    new_location_id = fields.Many2one(
+        comodel_name="stock.location",
+        string="New location",
+        required=True,
+        domain=lambda self: self._get_allowed_location_domain(),
+    )
+    moves_to_change = fields.Selection(
+        selection=[
+            ("all", "Change All moves"),
+            ("matched", "Change only moves with matched OLD location"),
+            ("manual", "Select moves to change"),
+        ],
+        string="Operations to change",
+        required=True,
+        default="manual",
+        help="Select which kind of selection of the moves you want to do",
+    )
+    move_lines = fields.Many2many("stock.move", string="Move lines")
+
+    def _check_allowed_pickings(self, pickings):
+        forbidden_pickings = pickings.filtered(
+            lambda x: x.state not in self._get_allowed_picking_states()
+        )
+        if forbidden_pickings:
+            raise UserError(
+                _(
+                    "You can not change operations destination location if "
+                    "picking state is not in %s"
+                )
+                % ",".join(self._get_allowed_picking_states())
+            )
+
+    def _check_allowed_moves(self, moves):
+        if any([move.move_orig_ids for move in moves]):
```

Comment:
Perhaps you should also check for allowed states in moves (as you also do in pickings). Sometimes you could have a cancelled move in an non cancelled picking

Reply:
You are right, added the extra check now :+1: